indexer/services/roundtable/src/tasks/aggregate-trading-rewards.ts

```diff
@@ -1,63 +1,230 @@
-
-  const tradingRewardData: TradingRewardFromDatabase[] = await getTradingRewardDataToProcess(
-    interval,
-  );
-  const sortedTradingRewardData: SortedTradingRewardData = sortTradingRewardData(tradingRewardData);
-  await updateTradingRewardsAggregation(sortedTradingRewardData);
-  // TODO(IND-499): Update AggregateTradingRewardsProcessedCache
+export default function generateTaskFromPeriod(
+  period: TradingRewardAggregationPeriod,
+): () => Promise<void> {
+  return async () => {
+    const aggregateTradingReward: AggregateTradingReward = new AggregateTradingReward(period);
+    await aggregateTradingReward.runTask();
+  };
 }
 
-async function getTradingRewardDataToProcessInterval(): Promise<Interval> {
-  const latestBlock: BlockFromDatabase = await BlockTable.getLatest();
+export class AggregateTradingReward {
+  period: TradingRewardAggregationPeriod;
 
-  // TODO(IND-499): Setup AggregateTradingRewardsProcessedCache for start time and add end time
-  return {
-    start: DateTime.fromISO(latestBlock.time),
-    end: DateTime.fromISO(latestBlock.time),
-  };
-}
+  constructor(period: TradingRewardAggregationPeriod) {
+    this.period = period;
+  }
 
-async function getTradingRewardDataToProcess(
-  _interval: Interval,
-): Promise<TradingRewardFromDatabase[]> {
-  // TODO: Implement
-  return Promise.resolve([]);
-}
+  async runTask(): Promise<void> {
+    const interval: Interval | undefined = await this.getTradingRewardDataToProcessInterval();
+    if (interval === undefined) {
+      logger.info({
+        at: 'aggregate-trading-rewards#runTask',
+        message: 'No interval to aggregate trading rewards',
+      });
+      return;
+    }
+    logger.info({
+      at: 'aggregate-trading-rewards#runTask',
+      message: 'Generated interval to aggregate trading rewards',
+      start: interval.start.toISO(),
+      end: interval.end.toISO(),
+    });
 
-function sortTradingRewardData(
-  _tradingRewardData: TradingRewardFromDatabase[],
-): SortedTradingRewardData {
-  // TODO: Implement
-  return {};
-}
+    const tradingRewardData:
+    TradingRewardFromDatabase[] = await this.getTradingRewardDataToProcess(interval);
+    const sortedTradingRewardData: SortedTradingRewardData = this.sortTradingRewardData(
+      tradingRewardData,
+    );
+    await this.updateTradingRewardsAggregation(sortedTradingRewardData);
+    await AggregateTradingRewardsProcessedCache.setProcessedTime(
+      this.period,
+      interval.end.toISO(),
+      redisClient,
+    );
+  }
+
+  /**
+   * Gets the interval of time to aggregate trading rewards for.
+   * If there are no blocks in the database, then do not process any data.
+   * If There is no processedTime in the cache, then delete the latest month of data,
+   * and reprocess that data or start from block 1.
+   * If the processedTime is not null, and blocks exist in the database, then process up to the
+   * next hour of data.
```

Comment:
why only process up the next hour of data and not the entire interval? regardless of interval, it should be single sql query to do aggregation right? are there latency concerns?

Reply:
I'm just worried about pulling too much data at once, I'll change the hour to a configurable variable